### ISSUE TITLE
statusChanges is not fired upon dynamic validation added with NG_VALIDATOR

### ISSUE THREAD

**Issue by waterplea** [angular/angular#38424]
# 🐞 bug report

### Affected Package

 The issue is caused by package @angular/forms

### Is this a regression?

Probably not

### Description

You can create directives that provide NG_VALIDATOR token so additional validation can be applied. Simple case, I have a checkbox that hides a particular field. If the field is enabled, I want it to be required. I can do this with built-in `required` directive. When I toggle my field, form becomes invalid because it is empty and required. However `statusChanges` stream does not fire to let me know that form became `INVALID`

## 🔬 Minimal Reproduction

https://stackblitz.com/edit/angular-ng-validator-bug

## 🌍  Your Environment

**Angular Version:**
<pre><code>10.0.1</code></pre>

**Anything else relevant?**
Watch out for ExpressionChanged error, form does actually become invalid the moment directive with NG_VALIDATOR is instantiated but the binding does not update. That's how I actually got to use statusChanges stream and noticed it's not firing.

**Comment by waterplea** [comment]
Also when you toggle it back to hidden validator is not removed and form stays `INVALID` but that might be by design.

**Comment by AndrewKushnir** [comment]
Hi @waterplea, thanks for reporting the issue and providing a repro.

It looks like in your example the Template-driven approach is mixed with the Reactive approach. I've created a [copy of your repro](https://stackblitz.com/edit/angular-ng-validator-bug-cjwl5r) and moved the `required` validator from the template to the `new FormControl` call and everything seems to be working correctly now (both form status as well as status change events). Please let me know if this solution works for you.

Thank you.

**Comment by waterplea** [comment]
Are you saying `NG_VALIDATOR` is only for template driven forms? It's not just about `required`, although it matches `formControlName` explicitly. It is a question for entire token, it was just easier to reproduce it with `required`. Also I completely got rid of `FormsModule` leaving only reactive forms and updated the example, same issue still present.

**Comment by AndrewKushnir** [comment]
> Are you saying NG_VALIDATOR is only for template driven forms?

No, I've just proposed a solution to unblock you.

> It's not just about required, although it matches formControlName explicitly. It is a question for entire token, it was just easier to reproduce it with required. Also I completely got rid of FormsModule leaving only reactive forms and updated the example, same issue still present.

Right, I think that there is a problem with Forms logic that requires further investigation.

**Comment by waterplea** [comment]
Got it, thanks for the quick reply. In case it's helpful, here's the actual case where I ran into this:
https://stackblitz.com/edit/angular-declarative-dynamic-validation?file=src%2Fapp%2Fvalidator%2Fvalidator.directive.ts&s=09

I was investigating declarative approaches to dynamic validation, as opposed to imperative `setValidators` calls. Since it's dynamic I had to call validation manually anyway, so it's not blocking, but still rather strange behaviour.
